Latency blog, installment #2: snoopi_deep and flamegraphs (#1120)
JuliaLang/www.julialang.org: blog/2021/01/snoopi_deep.md

- precompilation allows you to cache the results of type inference, thus saving time when you start using methods defined in the package
- caching becomes more effective when a higher proportion of calls are made with inferrable argument types (i.e., type-inference "succeeds"). Successful inference introduces links, in the form of `MethodInstance` backedges, that permit the entire call graph to be cached.

As a consequence, anyone interested in precompilation needs to pay attention to type inference: how much time does it account for, where is it spending its time, and what can be done to improve caching.
Julia itself provides the internal infrastructure needed to "spy" on inference, and the user-space utilities are in the [SnoopCompile] package.
Starting with Julia 1.2, it became possible to do a limited amount of "spying" on inference, and this infrastructure corresponds to SnoopCompile's `@snoopi` macro.
Julia 1.6 includes new changes that have permitted a far deeper look at what inference is doing.
Appropriately enough, SnoopCompile calls this `@snoopi_deep`.

The rich data collected by `@snoopi_deep` are useful for several different purposes.

> in this post, we'll describe the basic tool and show how it can be used to profile inference.

Reviewer: ```suggestion
In this post, we'll describe the basic tool and show how it can be used to profile inference.
```

Later posts will show other ways to use the data to reduce the amount of type-inference or cache its results.

## Collecting the data

To see `@snoopi_deep` in action, we'll use the following demo:

```
module SnoopDemo
    struct MyType{T} x::T end

    extract(y::MyType) = y.x

    function domath(x)
        y = x + x
        return y*x + 2*x + 5
    end

    dostuff(y) = domath(extract(y))

>     function packintype(x)
>         y = MyType(x)
>         return dostuff(y)
>     end

Reviewer: this name `packintype` was hard for me to parse, and i don't really know what it means.

Maybe you can call this `domathwithmytype(x)`?
Or since that's getting quite verbose, maybe `domath_with_mytype(x)`? I think i like that best.
```suggestion
    function domath_with_mytype(x)
        y = MyType(x)
        return dostuff(y)
    end
```

end

> ```

Reviewer: It might be nice in your demo to also include a type unstable function, to show the separate invocations to inference in the profile?

For example, when i was working on snoopi_deep, i usually used this example:
```julia
julia> module M
          i(x) = x+5
          h(a::Array) = i(a[1]::Integer) + 2
          g(y::Integer) = h(Any[y])
       end;
```
which has this cute little bump where you can see it coming _back_ to infer `i()` at _runtime_, as a result of dispatch:
<img width="800" alt="Screen Shot 2021-01-20 at 5 03 58 PM" src="https://user-images.githubusercontent.com/1582097/105246723-a8ec6a80-5b41-11eb-91b1-dba98c01bb2e.png">

It takes a bit more explanation to get at what's going on with the `::Integer` part  - but maybe you could comee up with a less contrived example of type instability..

Reviewer: (I see now that you reference this a bit later, and so maybe this would over-complicate things to include it this early... Up to you!) :)

Author: I want to keep it simple initially.

@@

This may not look like much, but there's a wealth of information hidden inside `tinf`.
While optional, before proceeding further it's recommended to check `tinf` for hints that [invalidation](https://julialang.org/blog/2020/08/invalidations/) may have influenced the result:

```
julia> staleinstances(tinf)
SnoopCompileCore.InferenceTiming[]
```

`staleinstances` extracts `MethodInstances` that have some "stale" (no longer callable) generated code.
In our case, it returns an empty list, meaning that it found no stale instances, which guarantees that no invalidation occurred.
There's nothing "funny" going on behind the scenes that will influence our results.

@@note
**Box 2** Unlike `@snoopr`, another macro exported by SnoopCompile, `staleinstances` does not distinguish between whether that stale code existed before you ran the `@snoopi_deep` block, or whether it got invalidated during the running of that block.
For detailed analysis of invalidations, `@snoopr` is the recommended tool.

> @@

Reviewer: 🔥 this is super cool! You've made these changes after I stopped paying attention. thanks for all of this; it's neat! :)

Reviewer 2: 😂 I see Nathan and I had similar sentiments!

   └─ InferenceTimingNode: 0.000122/0.000453 on InferenceFrameInfo for Main.SnoopDemo.dostuff(::MyType{Int64}) with 2 direct children
      ├─ InferenceTimingNode: 0.000083/0.000161 on InferenceFrameInfo for Main.SnoopDemo.extract(::MyType{Int64}) with 2 direct children
      │  ├─ InferenceTimingNode: 0.000045/0.000045 on InferenceFrameInfo for getproperty(::MyType{Int64}, ::Symbol) with 0 direct children
      │  └─ InferenceTimingNode: 0.000034/0.000034 on InferenceFrameInfo for getproperty(::MyType{Int64}, x::Symbol) with 0 direct children
      └─ InferenceTimingNode: 0.000170/0.000170 on InferenceFrameInfo for Main.SnoopDemo.domath(::Int64) with 0 direct children
```

This tree structure reveals the caller-callee relationships, showing the specific types that were used for each `MethodInstance`.
Indeed, as the calls to `getproperty` reveal, it goes beyond the types and even shows the results of [constant propagation](https://en.wikipedia.org/wiki/Constant_folding);
the `getproperty(::MyType{Int64}, x::Symbol)` (note `x::Symbol` instead of just plain `::Symbol`) means that the call was `getproperty(y, :x)`, which corresponds to `y.x` in the definition of `extract`.

> @@note
> **Box 3** Generally we speak of [call graphs](https://en.wikipedia.org/wiki/Call_graph) rather than call trees.
> But because inference results are cached (a.k.a., we only "visit" each node once), we obtain a tree as a depth-first-search of the full call graph.
> @@

Reviewer: ```suggestion
@@note
**Box 3** Generally we speak of [call graphs](https://en.wikipedia.org/wiki/Call_graph) rather than call trees.
But because inference results are cached (a.k.a., we only "visit" each node once), we obtain a tree as a depth-first-search of the full call graph.

Keep in mind that this can sometimes lead to surprising results when trying to improve inference times. You may delete a node in the graph, expecting to
eliminate it and all of its children, only to find that the children are now sitting under a _different_ parent node, the next function to be inferred that calls
your child, and was previously relying on it being cached.
@@
```

Each node in this tree is accompanied by a pair of numbers.
The first number is the *exclusive* inference time (in seconds), meaning the time spent inferring the particular MethodInstance, not including the time spent inferring its callees.
The second number is the *inclusive* time, which is the exclusive time plus the time spent on the callees.
Therefore, the inclusive time is always at least as large as the exclusive time.

The `ROOT` node is a bit different: its exclusive time measures the time spent on all operations *except* inference.
In this case, we see that the entire call took approximately 10ms, of which 9.3ms was spent on activities besides inference.
Almost all of that was code-generation, but it also includes the time needed to run the code.
Just 0.76ms was needed to run type-inference on this entire series of calls.
As you will quickly discover as you use `@snoopi_deep`, inference takes much more time on more complicated code.

> You can extract the `MethodInstance` with

Reviewer: ```suggestion
You can extract the `MethodInstance` that was being inferred at each node with
```
(might clarify this a bit for readers?)

```
julia> Core.MethodInstance(tinf)
MethodInstance for ROOT()

julia> Core.MethodInstance(tinf.children[1])
MethodInstance for packintype(::Int64)
```

## Visualizing the output

> We can also display this tree as a flame graph, using the [ProfileView](https://github.com/timholy/ProfileView.jl) package:

Reviewer: ```suggestion
We can also display this tree as a flame graph, using either the [ProfileView](https://github.com/timholy/ProfileView.jl) or [PProf](https://github.com/JuliaPerf/PProf.jl) packages:
```
😉 

We've been using _both_ PProf and ProfileView together when we've been investigating. ProfileView because it shows the profile in order, and PProf because it is built specifically for handling _very large_ profiles, and some of our profiles are... woof... huge:
<img width="1216" alt="Screen Shot 2020-12-18 at 9 58 40 AM" src="https://user-images.githubusercontent.com/1582097/105248075-cd494680-5b43-11eb-986b-978b7ce73c7d.png">

But the ability to search and filter and highlight and change through different views all together make it _super_ valuable for exploring very large codebases:
<img width="1439" alt="Screen Shot 2020-12-14 at 5 06 47 PM" src="https://user-images.githubusercontent.com/1582097/105248197-fbc72180-5b43-11eb-9485-b8d0c8d3be0c.png">

<img width="1552" alt="blurred2" src="https://user-images.githubusercontent.com/1582097/105249019-688eeb80-5b45-11eb-9c97-48c70fc879f2.png">
(hehe sorry i blurred it)

Maybe we can even include an example here of working with PProf and discuss the pros and cons of each tool? :)

Author: I'd be happy to have this mentioned! I'd rely on you to put that in, just because I don't have your experience.

- ctrl-click can be used to zoom in (you can do rubber band selection or zoom in/zoom out)
- empty horizontal spaces correspond to activities other than type-inference; in this case, the relatively narrow flame followed by a lot of empty space indicates that all the type inference happened at the beginning and accounted for only a small fraction of the total time.

You can explore this flamegraph and compare it to the output from `print_tree`.

In less trivial examples, these flame graphs may look more interesting:

![flamegraph-complex](/assets/blog/2021-latency/flamegraph-complex.png)

Here, the red boxes correspond to `MethodInstance`s which cannot be "naturally" precompiled.
This occurs when the method is owned by one module but the argument types are from another unrelated module.
We'll see how to deal with these in later installments.

You also see breaks in the flamegraph. During these periods, code generation and runtime create new objects and then call methods on those objects; if one of those calls requires a fresh entrance into inference, that triggers the creation of a new flame.

> Hence, the number of distinct flames (which is just equal to `length(tinf.children)`) gives you a rough indication of how frequently the chains of inference were broken.

Reviewer: ```suggestion
Hence, the number of distinct flames (which is just equal to `length(tinf.children)`) gives you a rough indication of how frequently the chains of inference were broken. This can be caused by type instability, separate top-level calls from the REPL during `@snoopi_deep`, calls to `eval` in your code, or some other cause for julia to start running type inference.
```

 InferenceTiming: 0.009382/0.010515 on InferenceFrameInfo for Core.Compiler.Timings.ROOT()
```

By default, this orders the nodes by exclusive time, but `flatten(tinf; sortby=inclusive)` allows you to sort by inclusive time.
Finally, `flatten(tinf; sortby=nothing)` returns the nodes in the order in which they were inferred.

Sometimes, you might infer the same method for many different specific argument types, and sometimes you might want to get a sense of the aggregate cost:

```
julia> accumulate_by_source(flatten(tinf))
8-element Vector{Tuple{Float64, Union{Method, Core.MethodInstance}}}:
 (7.838100000000001e-5, getproperty(x, f::Symbol) in Base at Base.jl:33)
 (8.2955e-5, extract(y::Main.SnoopDemo.MyType) in Main.SnoopDemo at REPL[1]:4)
 (0.000121738, dostuff(y) in Main.SnoopDemo at REPL[1]:11)
 (0.000132328, Main.SnoopDemo.MyType{T}(x) where T in Main.SnoopDemo at REPL[1]:2)
 (0.000170205, domath(x) in Main.SnoopDemo at REPL[1]:6)
 (0.000193107, Main.SnoopDemo.MyType(x::T) where T in Main.SnoopDemo at REPL[1]:2)
 (0.000354527, packintype(x) in Main.SnoopDemo at REPL[1]:13)
 (0.009381595, ROOT() in Core.Compiler.Timings at compiler/typeinfer.jl:75)
```

> This shows the cost of each `Method`, summing across all specific `MethodInstance`s.

Reviewer: We've found ourselves issuing queries like this, quite a bit:

```julia
julia> fs = [t for t in flattened if occursin("merge_sorted!", repr(t[2].mi.def.name));

julia> sum(first.(fs))  # what is the total time inferring all specializations of all methods of merge_sorted! 
341.62921501600005

julia> length(fs)  # how many specializations are there?
5968
```
Or using this to plot a distribution of the specialization times for this function:
<img width="480" alt="Screen Shot 2021-01-20 at 5 49 37 PM" src="https://user-images.githubusercontent.com/1582097/105250525-e94ee700-5b47-11eb-8927-dfefb674036c.png">

Reviewer: I'm not sure if it's worth mentioning things like this since we don't have standard built-in functions for things like this, and also hopefully most people are better off than we are, but 🤷 i thought i'd mention it :)

Author: `pgdsgui` will get at such issues, I favor leaving it for later.

To profile inference on this call, we simply do the following:

```
julia> using SnoopCompile

julia> tinf = @snoopi_deep SnoopDemo.domath_with_mytype(1)
InferenceTimingNode: 0.009382/0.010515 on InferenceFrameInfo for Core.Compiler.Timings.ROOT() with 1 direct children
```

@@note
**Box 1** Inference gets called only on the *first* invocation of a method with those specific types. You have to redefine the `SnoopDemo` module (by re-executing the command we used to define it) if you want to collect data with `@snoopi_deep` on the same code a second time.  This won't redefine any `Base` methods that `SnoopDemo` depends on, so you still may see differences between running this in a fresh session and subsequent uses of `@snoopi_deep`.
@@

This may not look like much, but there's a wealth of information hidden inside `tinf`.
While optional, before proceeding further it's recommended to check `tinf` for hints that [invalidation](https://julialang.org/blog/2020/08/invalidations/) may have influenced the result:

```
julia> staleinstances(tinf)
SnoopCompileCore.InferenceTiming[]
```

> `staleinstances` extracts `MethodInstances` that have some "stale" (no longer callable) generated code.

Reviewer 2: Wow nice!

      │  ├─ InferenceTimingNode: 0.000045/0.000045 on InferenceFrameInfo for getproperty(::MyType{Int64}, ::Symbol) with 0 direct children
      │  └─ InferenceTimingNode: 0.000034/0.000034 on InferenceFrameInfo for getproperty(::MyType{Int64}, x::Symbol) with 0 direct children
      └─ InferenceTimingNode: 0.000170/0.000170 on InferenceFrameInfo for Main.SnoopDemo.domath(::Int64) with 0 direct children
```

This tree structure reveals the caller-callee relationships, showing the specific types that were used for each `MethodInstance`.
Indeed, as the calls to `getproperty` reveal, it goes beyond the types and even shows the results of [constant propagation](https://en.wikipedia.org/wiki/Constant_folding);
the `getproperty(::MyType{Int64}, x::Symbol)` (note `x::Symbol` instead of just plain `::Symbol`) means that the call was `getproperty(y, :x)`, which corresponds to `y.x` in the definition of `extract`.

@@note
**Box 3** Generally we speak of [call graphs](https://en.wikipedia.org/wiki/Call_graph) rather than call trees.
But because inference results are cached (a.k.a., we only "visit" each node once), we obtain a tree as a depth-first-search of the full call graph.

> Keep in mind that this can sometimes lead to surprising results when trying to improve inference times. You may delete a node in the graph, expecting to

Reviewer 2: This pain: familiar to @NHDaly and I 😂.

Reviewer: haha yeah, i added that line... 😂

Author: He wrote that!

![PProf Flamegraph in full](/assets/blog/2021-latency/pprof-flamegraph-full-1.png)

Of course, as above, in this example inference was only a small part of the total time. You can zoom-in on nodes by clicking on them, and hovering will display their full text and time. The nicest way to expand just the inference time is to filter out the _non-inference time_ by "hiding" the `ROOT()` node. To do this, type `ROOT()` in the `Search regexp` bar and click REFINE -> Hide, or add `?h=ROOT` to the URL. This works well if there are several top-level calls to inference, as discussed below. One benefit of hiding and filtering via the Search bar is that the percentages for all nodes will be recomputed as a percentage of the new total.

Typing values into the search bar (without pressing enter) will highlight them in the graph, as displayed in the following image, and pressing enter will filter the graph to _only_ contain paths that match your search. These tools can be used to identify small functions that show up often, contributing large amounts of time when aggregated together.

![PProf Flamegraph filtered and highlighted](/assets/blog/2021-latency/pprof-flamepgraph-no-root-highlight.png)

> Here's a screenshot of the Top view, again with `ROOT()` hidden: `http://localhost:57599/ui/top?h=ROOT`

Author: ```suggestion
Here's a screenshot of the Top view, again with `ROOT()` hidden:
```

Reviewer: Yeah i'm not sure how i ended up with that punctuation. I wanted to just put a tiny reminder in there that you can see what's being hidden in the URL. Maybe shrinking it down like this would help?:
```suggestion
Here's a screenshot of the Top view, again with `ROOT()` hidden (`/top?h=ROOT()`):
```

> <img src="/assets/blog/2021-latency/pprof-top-no-root.png" width="600" title="PProf Top view" />

Author: This isn't displaying properly in the preview.

Reviewer: Hrmm.. bummer. it looked okay through the github page, when i looked here:
https://github.com/JuliaLang/www.julialang.org/blob/teh/inference_flamegraphs/blog/2021/01/snoopi_deep.md

Maybe I should just shrink down the size of the image first and then put the smaller image into the repo?

Author: That sounds expedient if there's no better solution.

Reviewer: Yeah, okay, this look good to go:
https://github.com/NHDaly/www.julialang.org/blob/nhd/inference_flamegraphs-pprof/blog/2021/01/snoopi_deep.md#pprofjl

I've created another PR. if you can merge it, it looks ready to me! :) thanks
https://github.com/JuliaLang/www.julialang.org/pull/1128
